windows/scripts/extract_kernels.py

```diff
@@ -0,0 +1,201 @@
+
+
+def parse_args():
+    """Parse the command-line arguments."""
+    parser = argparse.ArgumentParser(description='Extract Windows kernels.')
+    parser.add_argument('-d', '--iso-dir', required=True,
+                        help='Path to the directory containing the ISOs')
+    parser.add_argument('-o', '--output-dir', default=os.getcwd(),
+                        help='Directory to store the extract kernels. '
+                             'Defaults to the current working directory')
+
+    return parser.parse_args()
+
+
+def seven_zip_extract(source, files, dest=None):
+    """
+    Execute 7zip and extract the list of `files` from the `source` archive.
+    """
+    if not dest:
+        dest = os.getcwd()
+
+    args = ['7z', 'e', source]
+    args.extend(files)
+
+    proc = subprocess.Popen(args, stdout=DEVNULL, stderr=subprocess.PIPE,
+                            cwd=dest)
+    _, stderr = proc.communicate()
+
+    return proc.returncode, stderr
+
+
+def main():
+    """The main function."""
+    args = parse_args()
+
+    iso_dir = args.iso_dir
+    if not os.path.isdir(iso_dir):
+        raise Exception('%s is not a valid ISO directory' % iso_dir)
+
+    output_dir = args.output_dir
+    if not os.path.isdir(output_dir):
+        raise Exception('%s is not a valid output directory' % output_dir)
+
+    iso_dir = os.path.realpath(iso_dir)
+    output_dir = os.path.realpath(output_dir)
+
+    for iso, (container, krnl) in NTOSKRNL_MAP.items():
+        iso_path = os.path.join(iso_dir, iso)
+        if not os.path.isfile(iso_path):
+            print(u'[\u2717] %s does not exist. Skipping...' % iso)
+            continue
+
+        with TemporaryDirectory() as temp_dir:
+            print('[-] Extracting ntoskrnl.exe from %s...' % iso)
+
+            # Extract the kernel container
+            returncode, stderr = seven_zip_extract(iso_path, [container],
+                                                   temp_dir)
+            if stderr:
+                print(u'[\u2717] Failed to extract %s from %s: "%s"' %
+                      (container, iso, stderr))
+                continue
+
+            _, container_name = os.path.split(container)
+            container_path = os.path.join(temp_dir, container_name)
+
+            # Extract the kernel executable
+            returncode, stderr = seven_zip_extract(container_path, [krnl],
+                                                   temp_dir)
+            if returncode:
+                print(u'[\u2717] Failed to extract ntoskrnl.exe from %s: "%s"' %
+                      (container_name, stderr))
+                continue
+
+            # Rename the kernel executable (they are all called ntoskrnl.exe!)
+            # and move it to the output directory
+            iso_name, _ = os.path.splitext(iso)
+            shutil.move(os.path.join(temp_dir, 'ntoskrnl.exe'),
+                        os.path.join(output_dir, '%s_ntoskrnl.exe' % iso_name))
```

Comment:
You could also prefix the kernels with their hash (perhaps in addition to the image name).

Reply:
I think this relates to the previous comment. guest-images/s2e-env hard-codes the ISO file names (see https://github.com/S2E/guest-images/blob/master/images.json), so I did the same here for consistency.

If you are building an image using guest-tools then the ISO will already have this file name (otherwise the image won't build), so I just assumed it had the same filename

Reply:
I meant prefixing the kernels with the hash of the kernel (not the ISO). This would solve the problem of potentially having two kernels with identical names on the same image but different hashes.

Reply:
Ah ok, I understand. Makes sense!